### ISSUE TITLE
errors running terraform destroy, unable to debug

### ISSUE THREAD

**Issue by ketzacoatl** [hashicorp/terraform#3743]
I have refactored a few of my Terraform modules, and I am now running into a `Error creating plan: 1 error(s) occurred: * Cycle: ...` error which has been difficult to debug.

I have a set of Terraform modules for deploying consul leader/agent clusters to AWS auto-scaling groups, taking care of networks, keys, security groups, init/provisioning, the full bit. I reuse the agent cluster module to create new types of auto-scaling clusters (with consul at the heart of the cluster). My initial MVP functioned best when I included everything (network, key pair, security groups) in with each of the agent/leader modules, and I ran into issues with nested modules and template files, so I kept things simpler for the MVP.

I am now at a point where I would like to simplify the modules and remove redundancy. EG, some of the resources can be shared across clusters, but I need to shuffle around the resources to do so. Having completed some of this refactor, I am now testing the result with Terraform more thoroughly, and I am running into the following error when I try to destroy the whole thing:

```
Error creating plan: 1 error(s) occurred:

* Cycle: module.cworkers-a.module.agent-asg.output.lc_id, module.cworkers-a.module.cluster-net.output.id_a, module.cworkers-a.module.cluster-net.aws_subnet.a, module.cworkers-a.module.agent-asg.var.subnet_ids, module.cworkers-a.var.vpc_id, module.cworkers-a.module.cluster-net.var.vpc_id, module.cworkers-a.module.cluster-net.aws_subnet.c, module.cworkers-a.module.cluster-net.output.id_c, module.cworkers-a.module.agent-asg.aws_autoscaling_group.cluster, module.cworkers-a.var.ssh_security_group, module.cworkers-a.module.agent-asg.var.security_group_ids, module.test-vpc.output.cidr_block, module.consul-agent-sg.var.cidr_blocks, module.consul-agent-sg.var.vpc_id, module.consul-agent-sg.aws_security_group.main, module.consul-agent-sg.output.id, module.cworkers-a.var.agent_security_group, module.cworkers-a.module.agent-asg.var.user_data, module.cworkers-a.module.agent-asg.aws_key_pair.cluster (destroy), module.cworkers-a.module.agent-asg.aws_key_pair.cluster, module.cleaders.var.vpc_id, module.cleaders.aws_route53_zone.leaders, module.cleaders.aws_route53_record.leaders, module.cleaders.output.leader_dns, template_file.worker-init, module.cworkers-a.var.user_data, aws_security_group.worker-service, module.cworkers-a.var.service_security_group, module.consul-agent-sg.aws_security_group.main (destroy), template_file.worker-init (destroy), module.cleaders.aws_route53_record.leaders (destroy), module.cleaders.aws_route53_zone.leaders (destroy), module.test-vpc.aws_vpc.core (destroy), module.test-vpc.aws_vpc.core, module.test-vpc.output.id, module.public-ssh-sg.var.vpc_id, module.public-ssh-sg.aws_security_group.ssh (destroy), module.public-ssh-sg.aws_security_group.ssh, module.public-ssh-sg.output.id, module.cworkers-a.module.agent-asg.aws_launch_configuration.cluster, module.cworkers-a.module.agent-asg.output.lc_name, module.cworkers-a.module.agent-asg.aws_launch_configuration.cluster (destroy), aws_security_group.worker-service (destroy)
```

I have a difficult time making sense out of the error message, other than _there must be a cyclical dependency issue with my code_. Here is why I am confused:
* the code is relatively simple, and I'm not seeing where/how TF is getting tripped up
* TF can create the environment just fine, but when I run `tf destroy -force`, I get the `Cycle` error.
* Interestingly, if I `tf destroy -force -target=aws_key_pair.tests`, then `tf destroy -force`, there is no problem/error in either run, everything is destroyed
* when I use `tf graph > g.dot` and try to open with `xdot`, that fails. I also tried using the `-draw-cycles` flag, but I am not familiar with dot and graphiz and maybe I should use a different app to review the graph?

The code is not public (yet, I just spent some time to separate out client/proprietary from generic open-source), but I can get TF developers into the repo to review and test things out.

**Comment by apparentlymart** [comment]
Sorry for all of this trouble. Are you using the `create_before_destroy` lifecycle flag for any of your resources? The most common cause of cycles is mixing resources that have that flag with resources that don't, as was documented in #2944:

> **NOTE on create\_before\_destroy and dependencies:** Resources that utilize the `create_before_destroy` key can only depend on other resources that also include `create_before_destroy`. Referencing a resource that does not include `create_before_destroy` will result in a dependency graph cycle. 

If that doesn't seem like your issue, then I have some hints for `terraform graph`: adding the `-verbose` flag will make the graph include some additional nodes that are normally hidden to keep the graph simple; unfortunately some cycles -- including the one with `create_before_destroy` above -- are only visible in `-verbose` mode.

If you run `terraform graph -verbose -draw-cycles`

**Comment by ketzacoatl** [comment]
You are probably right - I likely have `create_before_destroy` inconsistently. Thank you for the info @apparentlymart, I will dig through the modules to find the one or two sprinkled in there, and report back.

A question in the meantime.. can you put `create_before_destroy` on a module, is it even safe/sane/sensible to do so?

Also, what is the recommended way to view the .dot output from `tf graph`?

**Comment by ketzacoatl** [comment]
You were correct @apparentlymart, I found a resource with `create_before_destroy` in one of the modules reused a few times, and dropping this lifecycle block lets me create/destroy in one go without cycle errors. THANK YOU! Also, I found the dot to png conversion in https://terraform.io/docs/commands/graph.html, but viewing the image crashed my host. I will see what I can do about that on my own.

I
